client/src/components/ExpandedRowComponent.tsx

```diff
@@ -0,0 +1,128 @@
+import React, {MouseEvent} from 'react';
+import type {CodeReview} from '../lib/types';
+import {format} from 'date-fns';
+import {logError} from '../services/log';
+import {useState} from 'react';
+import {fetchData} from '../services/fetch';
+
+export const useExpandedRowComponent = ({onUpdate, user}: {onUpdate: CallableFunction; user?: {displayName: string}}) => {
+    const ExpandedRowComponent = ({data}: {data: CodeReview}) => {
+        const handleActionClick = ({currentTarget}: MouseEvent<HTMLButtonElement>) => {
+            const action = currentTarget.getAttribute('name');
+            const value = currentTarget.getAttribute('value');
+
+            fetchData(`/api/action/${action}/${value}`)
+                .then((result) => {
+                    result?.data && onUpdate && onUpdate({
+                        data: result.data,
+                        action,
+                    });
+                })
+                .catch((e) => {
+                    setErrorMessage('Error saving data, see console error log for details.');
+                    logError(`Error sending: /api/action/${action}/${value}`, e);
+                })
+        };
+        const ActionButton = ({label, reviewId}: {label: string; reviewId: number}) => {
+            const sanitizedName = label.toLowerCase().replace(' ', '-');
+
+            return (
+                <button
+                    id={`${sanitizedName}-${reviewId}`}
+                    name={sanitizedName}
+                    value={reviewId}
+                    onClick={handleActionClick}
+                >{label}</button>
+            );
+        };
+        const showClaim = !user || !data.reviewers?.includes(user?.displayName);
+        const showApprove = !user || !data.approvers?.includes(user?.displayName);
+        const showChange = !user || !data.requestChanges?.includes(user?.displayName);
+        const [editing, setEditing] = useState(false);
+        const [errorMessage, setErrorMessage] = useState(null as null | string);
+        const [saveClicked, setSaveClicked] = useState(false);
+        const dirtyData = {...data};
```

Comment:
This seems like a code smell too - you shouldn't need to keep track of oldProps vs newProps in the component.

Reply:
This was use to add support of row edit to keep track of data changes.  When Save button is call, this data will get all updated and send to the server.  We probably can useRef for this.  Decide to use track the entire object individual fields: note, jiraTicket, etc... more in the future?

Alternative is to use un-controlled edit form and manually collect those data when Save button is clicked.